File: js_backup/agent.js
import { allTools } from "./tools/index.js";
import { CONFIG } from "./config.js";
import { getIntent } from "./router.js";
import { formatToolResult } from "./formatter.js";
import { chat } from "./llm.js";

/* ---------------- TOOL EXECUTION ---------------- */

export async function executeTool(toolName, args) {
    const toolDefinition = allTools.find(t => t.name === toolName);

    if (!toolDefinition) {
        return { status: "error", data: null, message: `Unknown tool: ${toolName}` };
    }

    console.log(`[Execution] Invoking tool: ${toolDefinition.name}`);
    console.log(`[Execution] Args: ${JSON.stringify(args)}`);

    try {
        const result = await toolDefinition.execute(args);

        // --- 7. TOOL EXECUTION VALIDATION ---
        const isValid = toolDefinition.validationRule(result);
        console.log(`[Execution] Tool Result Validity: ${isValid ? "VALID" : "INVALID"}`);

        return { result, isValid };
    } catch (err) {
        console.error(`[Execution] Error in ${toolDefinition.name}:`, err.message);
        return { result: { status: "error", data: null, message: err.message }, isValid: false };
    }
}

/* ---------------- MAIN AGENT ---------------- */

export async function runAgent(query, chatHistory, isInteractive = false) {
    if (!query || !query.trim()) {
        console.log("[Agent] Empty query received. Skipping.");
        return;
    }
    const startTime = performance.now();
    console.log("\n" + "=".repeat(50));
    console.log(`[Agent] Processing Query: "${query}"`);
    console.log("=".repeat(50));

    // 1. Pre-processing Timer
    const preRAGStart = performance.now();
    const intent = await getIntent(query);
    const preRAGTime = (performance.now() - preRAGStart).toFixed(1);

    let finalAnswer = "";

    if (intent.path === "DIRECT_TOOL") {
        const { result, isValid } = await executeTool(intent.tool, intent.args || {});
        if (isValid) {
            finalAnswer = formatToolResult(query, result);
        } else {
            intent.path = "RAG_FALLBACK";
        }
    }

    if (intent.path === "RAG_FALLBACK" || intent.path === "PHI:latest") {
        console.log("\n[Decision Path] Entering RAG Fallback...");

        const ragTool = allTools.find(t => t.name === "search_knowledge_rag");

        // RAG Timing Breakdown
        const embStart = performance.now();
        const ragRes = await ragTool.execute({ query }); // Logic handles embedding + search
        const totalRAGTime = (performance.now() - embStart).toFixed(1);

        console.log(`Confidence Level: ${ragRes.confidence}`);
        console.log(`[RAG Metrics] Max: ${ragRes.scores?.max?.toFixed(2) || 0} | Avg: ${ragRes.scores?.avg?.toFixed(2) || 0}`);

        const genStart = performance.now();
        if (ragRes.confidence === "HIGH") {
            console.log("[Decision Path] RAG Confidence HIGH -> Phi Grounded Mode");
            const prompt = `Answer in one short sentence (<15 words) using context: ${ragRes.data}\n\nQuestion: ${query}`;
            const response = await chat([{ role: "user", content: prompt }], null, { num_predict: 30 });
            finalAnswer = response.content;
        } else if (ragRes.confidence === "MEDIUM") {
            console.log("[Decision Path] RAG Confidence MEDIUM -> Phi Grounded Mode (Aggressive Limit)");
            // MEDIUM RULE: limit to 1 chunk, enforce very short output
            const prompt = `Context: ${ragRes.data}\n\nQuestion: ${query}\nInstruction: Return only examples or a direct answer. Max 10 words.`;
            const response = await chat([{ role: "user", content: prompt }], null, { num_predict: 20 });
            finalAnswer = response.content;
        } else {
            console.log("[Decision Path] RAG Confidence LOW -> Phi General Mode");
            const prompt = `Answer in 2-3 detailed sentences (~50 words): ${query}`;
            const response = await chat([{ role: "user", content: prompt }], null, { num_predict: 100 });
            finalAnswer = response.content;
        }
        const genTime = (performance.now() - genStart).toFixed(1);

        // HARD OUTPUT TRIM (Allow more sentences in General Mode)
        if (ragRes.confidence !== "LOW") {
            finalAnswer = finalAnswer.split(/[.!?]/)[0] + ".";
        } else {
            // In general mode, just ensure it's not excessively long but allow multiple sentences
            finalAnswer = finalAnswer.trim();
        }

        // Performance Telemetry
        console.log(`\n[Telemetry] Pre-processing: ${preRAGTime}ms`);
        console.log(`[Telemetry] RAG (Emb+Search): ${totalRAGTime}ms`);
        console.log(`[Telemetry] Phi Generation: ${genTime}ms`);
        console.log(`[Telemetry] Total Latency: ${(performance.now() - startTime).toFixed(1)}ms`);
    }

    console.log("\n" + "-".repeat(20) + " FINAL ANSWER " + "-".repeat(20));
    console.log(finalAnswer);
    console.log("-".repeat(50));

    chatHistory.push({ role: "user", content: query });
    chatHistory.push({ role: "assistant", content: finalAnswer });
}

File: server/agent.py
import time
from typing import List, Dict, Any, Tuple
from .tools import all_tools
from .config import CONFIG
from .router import get_intent
from .formatter import format_tool_result
from .llm import chat

def execute_tool(tool_name: str, args: Dict[str, Any]) -> Tuple[Dict[str, Any], bool]:
    tool_def = next((t for t in all_tools if t.name == tool_name), None)
    
    if not tool_def:
        return {"status": "error", "data": None, "message": f"Unknown tool: {tool_name}"}, False
        
    print(f"[Execution] Invoking tool: {tool_def.name}")
    print(f"[Execution] Args: {args}")
    
    try:
        result = tool_def.execute(**args)
        
        is_valid = tool_def.validation_rule(result)
        print(f"[Execution] Tool Result Validity: {'VALID' if is_valid else 'INVALID'}")
        
        return result, is_valid
    except Exception as e:
        print(f"[Execution] Error in {tool_def.name}: {str(e)}")
        return {"status": "error", "data": None, "message": str(e)}, False

def run_agent(query: str, chat_history: List[Dict[str, str]], is_interactive: bool = False):
    if not query or not query.strip():
        print("[Agent] Empty query received. Skipping.")
        return
        
    start_time = time.time()
    print("\n" + "=" * 50)
    print(f"[Agent] Processing Query: \"{query}\"")
    print("=" * 50)
    
    pre_rag_start = time.time()
    intent = get_intent(query)
    pre_rag_time = (time.time() - pre_rag_start) * 1000
    
    final_answer = ""
    
    if intent["path"] == "DIRECT_TOOL":
        result, is_valid = execute_tool(intent["tool"], intent.get("args", {}))
        if is_valid:
            final_answer = format_tool_result(query, result)
        else:
            intent["path"] = "RAG_FALLBACK"
            
    if intent["path"] in ["RAG_FALLBACK", "PHI:latest"]:
        print("\n[Decision Path] Entering RAG Fallback...")
        
        rag_tool = next((t for t in all_tools if t.name == "search_knowledge_rag"), None)
        
        emb_start = time.time()
        rag_res = rag_tool.execute(query=query) if rag_tool else {"confidence": "LOW", "data": ""}
        total_rag_time = (time.time() - emb_start) * 1000
        
        confidence = rag_res.get("confidence", "LOW")
        scores = rag_res.get("scores", {"max": 0, "avg": 0})
        
        print(f"Confidence Level: {confidence}")
        print(f"[RAG Metrics] Max: {scores.get('max', 0):.2f} | Avg: {scores.get('avg', 0):.2f}")
        
        gen_start = time.time()
        if confidence == "HIGH":
            print("[Decision Path] RAG Confidence HIGH -> Phi Grounded Mode")
            prompt = f"Answer in one short sentence (<15 words) using context: {rag_res.get('data')}\n\nQuestion: {query}"
            response = chat([{"role": "user", "content": prompt}], options={"num_predict": 30})
            final_answer = response.get("content", "")
        elif confidence == "MEDIUM":
            print("[Decision Path] RAG Confidence MEDIUM -> Phi Grounded Mode (Aggressive Limit)")
            prompt = f"Context: {rag_res.get('data')}\n\nQuestion: {query}\nInstruction: Return only examples or a direct answer. Max 10 words."
            response = chat([{"role": "user", "content": prompt}], options={"num_predict": 20})
            final_answer = response.get("content", "")
        else:
            print("[Decision Path] RAG Confidence LOW -> Phi General Mode")
            prompt = f"Answer in 2-3 detailed sentences (~50 words): {query}"
            response = chat([{"role": "user", "content": prompt}], options={"num_predict": 100})
            final_answer = response.get("content", "")
            
        gen_time = (time.time() - gen_start) * 1000
        
        if confidence != "LOW":
            parts = __import__('re').split(r'[.!?]', final_answer)
            final_answer = parts[0] + "." if parts else final_answer
        else:
            final_answer = final_answer.strip()
            
        print(f"\n[Telemetry] Pre-processing: {pre_rag_time:.1f}ms")
        print(f"[Telemetry] RAG (Emb+Search): {total_rag_time:.1f}ms")
        print(f"[Telemetry] Phi Generation: {gen_time:.1f}ms")
        print(f"[Telemetry] Total Latency: {(time.time() - start_time) * 1000:.1f}ms")
        
    print("\n" + "-" * 20 + " FINAL ANSWER " + "-" * 20)
    print(final_answer)
    print("-" * 50)
    
    chat_history.append({"role": "user", "content": query})
    chat_history.append({"role": "assistant", "content": final_answer})
